drivers/meter/ade7758/ade7758.c

```diff
@@ -0,0 +1,958 @@
+	/* value read from register */
+	uint32_t reg_val;
+	/* timeout value */
+	uint16_t timeout = 1000;
+
+	dev = (struct ade7758_dev *)no_os_calloc(1, sizeof(*dev));
+	if (!dev)
+		return -ENOMEM;
+
+	/* SPI Initialization */
+	ret = no_os_spi_init(&dev->spi_desc,
+			     init_param.spi_init);
+	if (ret)
+		goto error_spi;
+
+	/* interrupt */
+	dev->irq_ctrl = init_param.irq_ctrl;
+	/* software reset */
+	ret = ade7758_sw_reset(dev);
+	if (ret)
+		goto error_spi;
+	/* wait for reset ack */
+	do {
+		ret = ade7758_get_int_status(dev, ADE7758_RESET_MSK, &reg_val);
+		if (ret)
+			goto error_spi;
+		timeout--;
+	} while ((!reg_val) && (timeout > 0));
+	if (timeout <= 0) {
+		ret = -ETIMEDOUT;
+		goto error_spi;
+	}
+	/* reset the status register */
+	ret = ade7758_clear_irq_status(dev, &reg_val);
+	if (ret)
+		goto error_spi;
+	/* Read version product */
+	ret = ade7758_version_product(dev, &reg_val);
+	if (ret)
+		goto error_spi;
+
+	if (reg_val != ADE7758_VERSION) {
+		pr_err("ERROR device not recognized!!!\n");
+		goto error_spi;
+	}
+
+	*device = dev;
+
+	return 0;
+
+error_spi:
+	no_os_spi_remove(dev->spi_desc);
+error_dev:
+	no_os_free(dev);
+
+	return ret;
+}
+
+/**
+ * @brief Setup the ADE7758 device.
+ * @param dev - The device structure.
+ * @return	0 in case of success, negative error code otherwise.
+ */
+int ade7758_setup(struct ade7758_dev *dev)
+{
+	int ret;
+	/* register value */
+	uint32_t reg_val;
+
+	/* set the waveform register update rate */
+	ret = ade7758_wave_update_rate(dev, ADE7758_DATA_RATE_26_04KSPS);
+	if (ret)
+		return ret;
+
+	/* set the source of sampled data for wave register */
+	ret = ade7758_wave_sample_data_source(dev, ADE7758_VOLTAGE);
+	if (ret)
+		return ret;
+
+	/* set the phase of sampled data for wave register */
+	ret = ade7758_wave_phase_sel(dev, ADE7758_PHASE_A);
+	if (ret)
+		return ret;
+
+	/* set the input to the energy accumulation registers */
+	ret = ade7758_consel(dev, ADE7758_CONSEL_0);
+	if (ret)
+		return ret;
+
+	/* set the source of the measurement of the voltage line frequency */
+	ret = ade7758_freq_source(dev, ADE7758_PHASE_A);
+	if (ret)
+		return ret;
+
+	/* set the operation mode of the ADCs */
+	ret = ade7758_adcs_mode(dev, ADE7758_NORMAL_OPERATION);
+	if (ret)
+		return ret;
+
+	/* enable zero crossing interrupts */
+	ret = ade7758_enable_irq(dev, ADE7758_ZXA_MSK, ENABLE);
+	if (ret)
+		return ret;
+	ret = ade7758_enable_irq(dev, ADE7758_ZXB_MSK, ENABLE);
+	if (ret)
+		return ret;
+	ret = ade7758_enable_irq(dev, ADE7758_ZXC_MSK, ENABLE);
+	if (ret)
+		return ret;
+
+	return 0;
+}
+
+/**
+ * @brief Read device register.
+ * @param dev - The device structure.
+ * @param reg_addr - The register address.
+ * @param reg_data - The data read from 32bit register.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ade7758_read(struct ade7758_dev *dev, uint16_t reg_addr,
+		 int32_t *reg_data)
+{
+	int ret;
+	/* number of bytes = 2 for 8 bits register data */
```

Comment:
It's up to you if you want to keep them, but these comments aren't really necessary